### ISSUE TITLE
Circular dependencies for ESM modules

### ISSUE THREAD

**Issue by lorisleiva** [metaplex-foundation/solita#22]
## Problem

When compiling the beet structures auto-generated by Solita, there is a high chance of getting circular dependencies.

Here's an example when compiling the `tokenMetadata` generated modules.

<img width="1728" alt="CleanShot 2022-03-07 at 11 03 05@2x" src="https://user-images.githubusercontent.com/3642397/157019233-4a43a09d-c073-41bb-b7c3-9d4e2822cf41.png">

The issue comes from the fact that all `types` modules are importing the following line:

```ts
import * as definedTypes from '.';
```

This creates a circular dependency because the module is itself part of this `index.js` file.

Circular dependencies are usually fine in ESM module because modules usually export deferred variables such as classes and functions. Therefore, they are clever enough to know how to import a module that doesn't need to be executed right away.

However, with beet, all structures are exported as constants which creates the following runtime error when trying to execute the compiled code.

<img width="690" alt="CleanShot 2022-03-07 at 10 31 37@2x" src="https://user-images.githubusercontent.com/3642397/157020088-49f3a2c9-f39e-459b-a9a1-defb0d632b89.png">

In this example, the `createMetadataAccountArgsBeet` constant requires `definedTypes.dataBeet` which has not yet been initialised because `CreateMetadataAccountArgs.ts` is imported before `Data.ts` in the `index.ts` file (exported by alphabetical order).

In a typical app, we'd simply need to order the exports in a careful order but because these are auto-generated, it is not really an option for us. We could be clever about it and find the cycles in Solita before generating the modules but this feels like an overkill for me.

## Solution

When doing a bit of googling, I found [the following SO answer](https://stackoverflow.com/a/46593566/11440277).

> What's the best practice about circular dependencies in ES6?
>
> Avoid them altogether. If you cannot (or don't want to) completely avoid them, restrict the involved modules to only use function declarations with circular dependencies, never initialise top-level values (constants, variables, classes) by using imported values (that includes extends references).

So I think the easiest way for us to go around this is to change the API generated by Solita so it doesn't export structures directly but methods that return that structure.

For instance:

```diff
- export const createMasterEditionArgsBeet = new beet.FixableBeetArgsStruct<CreateMasterEditionArgs>(
+ export const createMasterEditionArgsBeet = () => new beet.FixableBeetArgsStruct<CreateMasterEditionArgs>(
    [['maxSupply', beet.coption(beet.u64)]],
    'CreateMasterEditionArgs',
  );
```

@thlorenz What do you think?

**EDIT**

Another (and potentially better) solution would be to update the Solita logic so they explicitly import what they need instead of requiring the entire folder. For example:

```diff
- import * as definedTypes from '.';
+ import { DataV2, dataV2Beet } from './DataV2';
```

This requires Solita to know which other types the current type depends on but I'm sure it already knows that in order to generate the file.

**Comment by thlorenz** [comment]
I already ran into this for a contract I'm doing, so this is not only a ESM related issue.

- [ ] referencing `*` of `definedTypes` causes circular imports
  - _fixed_ [here](https://github.com/metaplex-foundation/metaplex-program-library/commit/12b36b47a53d7ef895f71560feadf77abc62879b)

It's on my list to fix, but I was thinking of a solution that avoids `import * from <some-index>` and turn that into and `import { WhatINeed } from <exact-file>` instead.

What you propose would work as well, but not for importing types/object hashes and so on.

**Comment by lorisleiva** [comment]
Nice! I think this is a better solution because it also removes all circular dependency warning from the console when compiling using certain bundlers that check for it. 🙂

**Comment by thlorenz** [comment]
Fixed via #23.
